bsip-0040.md

```diff
@@ -0,0 +1,322 @@
+
+#### `logical_or`
+Stateless assert, only for dictionary type objects. The data is a list of restrictions lists, i.e. 
+`data = [ [restriction 1, restriction 2], [restriction 3, restriction 4], ... ]`. If one of the restrictions sub-lists in data passes as whole, this restriction passes. In above miniexample that would mean if `restriction 1` and `restriction 2` pass, the whole `logical_or` restriction is considered to be passed as well.
+
+### Outline of handling incoming transactions
+ 
+When a signed transaction arrives and before the backend evaluates if all necessary authorities are present through the signatures, do the following:
+- iterate over required accounts and for each account, iterate over all operations within the transactions that require the active authority of this account
+- iterate the `custom_active_authorities` of said account
+- if a `custom_active_authority` is found that matches , remember that and stop iterating the authorities and continue until all operations are checked
+- if the account has a `custom active authority` match for every operation in the transaction that requires it, then grant the `active authority` of said account. If no match is found, treat as if no authority was given
+
+Note:
+- A `custom_active_authority` can only grant the `active authority` of the corresponding account, nothing more
+
+### Economics
+
+Adding a custom active authority means increased effort for the backend, and with a stateful one also the need for more storage. Proposed transaction fees:
+- `install_custom_active_authority`: Tied to the duration of the custom active authority. 
+Normal accounts can only create custom active authoritites with a duration of maximum 1 year. LTM can do any duration and also unlimited, but the transaction fee is capped at duration of 2 years.
+- `update_custom_active_authority`: Base fee similar to `account_update` plus dynamic one for any duration changes
+- `delete_custom_active_authority`: Cheap similar to `limit_order_cancel`
+
+### Modification to the backend
+
+* Add a new index or extend the account object to store custom active permission are assigned to an account and contain a list of custom active authorities. Multiple custom active authority entries are possible for one operation
+* If the active authority of the account is updated, all custom active authorities need to be confirmed in the update. Every unconfirmed one is deleted otherwise
+* Provide operations: `install_custom_active_authority`, `update_custom_active_authority`, `delete_custom_active_authority` to allow changing the custom active permission (3 operation to allow custom transaction fees and avoid having to send the complete list of all authorities for every update)
```

Comment:
Please clarify: is it possible to set custom active permissions in an account_update operation? Or must an account_update operation be bundled together with an install_custom_active_authority operation in order to preserve the existing custom active authorities?

Reply:
The idea is to keep active authory as is, with power to do everything except chain owner key.
Additionally to that, additional custom_operation_permission can be installed that are linked to account-id in a `M to N` manner. All custom_operation_permissions would get their own "id" to be able to edit or delete them.

Reply:
The idea is that the user can change active key without having to pay the install fee again (because that should be somewhat expensive).

Best would be to keep the three new operations for custom active permission.

Cases:
1. bundle: account_update and update_custom_active in one transaction, where update_custom only says: be valid for this new active authority
2. implicit with account_update. Include a new array with reference to all custom_active_authorities that should remain
3. dont allow it at all. account_update deletes all custom_active_authorities, always

Which technical possibilities do we have roughly @pmconrad ?

Reply:
My question above aims at a contradiction between the previous two lines: account_update must allow specifying custom active authorities because otherwise they'll be deleted, OTOH you introduce new operations for this because you want dedicated fees.

Reply:
Having thought about it I propose:

A custom active authority has a status that can be enabled or disabled. This status can be set by the `update_custom_active_authority`. Additionally, if active authority is changed, all `update_custom_active_authority`s become disabled ALWAYS. User can enable it again with seperate call afterwards.

Reasoning:
 - avoid any misuse when accounts are given to people. Example: Account A is created by Person A and given to Person B. Person B changes all keys in order to fully own the account. But Person A has put a custom active authority, which Person B doesnt know and thus gains access
 - user should not be forced to pay the fee of `install_custom_active_authority` (expensive) again if he updates his active authority
 - it is acceptable to force the user to enable the custom active authorities after changing active keys

Reply:
Technically, we can add a field in `extensions` of `account_update_operation`, to specify a list of custom authorities to keep or remove.

Reply:
I included that option now